File: app/src/lib/api/mock-multipad.ts
// The mock's Multi Pads: what the engine does (docs/multipad.md), closely enough to build
// and screenshot the Multi Pad panel. No audio. MockSession calls in on its commands, bar
// lines, band start/stop, Ending starts and chords.

import type { MultiPadCmd, MultiPadState, PadLamp } from './types'

const ROOT = '/Users/me/Styles'

/** The mock's bank files: the synthetic demo bank (`yahaha pad --demo`) and two more. */
const BANKS = [
  { name: 'Demo', folder: 'Pads', pads: ['Shaker Loop', 'Rise Arp', 'Bass Riff', 'Brass Hit'], repeat: [true, false, true, false], cm: [false, true, true, true] },
  { name: 'Strings FX', folder: 'Pads', pads: ['Swell', 'Pizz Run', 'Stab', 'Tremolo'], repeat: [false, false, false, true], cm: [true, true, true, true] },
  { name: 'Latin Perc', folder: 'Pads/Latin', pads: ['Conga Loop', 'Timbale Fill', 'Cowbell', ''], repeat: [true, false, true, false], cm: [false, false, false, false] },
]

export function initialMultiPad(): MultiPadState {
  return {
    bank: null,
    loading: false,
    pads: [0, 1, 2, 3].map((i) => ({ index: i, name: '', lamp: 'empty' as PadLamp, repeat: false, chordMatch: false, channel: 5 + i })),
    synchroStop: { styleStop: true, ending: false },
    banks: BANKS.map((b, id) => ({ id, name: b.name, folder: b.folder, path: `${ROOT}/${b.folder}/${b.name}.pad` })),
  }
}

export class MockPads {
  /** Beats left of each playing pad's pass (every mock pad is one 4-beat bar). */
  private left = [0, 0, 0, 0]

  constructor(private get: () => MultiPadState) {}

  private get st() {
    return this.get()
  }

  private setLamp(i: number, lamp: PadLamp) {
    this.st.pads[i].lamp = lamp
    if (lamp === 'playing') this.left[i] = 4
  }

  private has(i: number) {
    const p = this.st.pads[i]
    return p !== undefined && p.lamp !== 'empty'
  }

  /** A press or a Synchro Start: at once when stopped, at the next bar line while playing. */
  private start(i: number, running: boolean) {
    if (!this.has(i)) return
    this.setLamp(i, running ? 'queued' : 'playing')
  }

  /** A pad command. Returns an error message for a refused one. */
  cmd(c: MultiPadCmd, running: boolean): string | null {
    const st = this.st
    const pad = 'pad' in c ? c.pad : 0
    if ('pad' in c && (pad < 0 || pad > 3)) return `no Multi Pad ${pad} (pads are 0-3)`
    switch (c.type) {
      case 'loadMultiPad':
      case 'loadMultiPadPath': {
        const id = c.type === 'loadMultiPad' ? c.id : st.banks.findIndex((b) => b.path === c.path)
        const b = BANKS[id]
        if (!b) return c.type === 'loadMultiPad' ? `no Multi Pad bank ${c.id}` : `${c.path}: not a MIDI/Multi Pad file`
        st.bank = { id, name: b.name, path: st.banks[id].path }
        st.pads = b.pads.map((name, i) => ({
          index: i, name, lamp: (name ? 'ready' : 'empty') as PadLamp, repeat: name ? b.repeat[i] : false,
          chordMatch: name ? b.cm[i] : false, channel: 5 + i,
        }))
        break
      }
      case 'clearMultiPad':
        st.bank = null
        st.pads = initialMultiPad().pads
        break
      case 'triggerMultiPad':
        // Pressing any pad starts the pads in standby with it.
        st.pads.forEach((p, i) => p.lamp === 'armed' && this.start(i, running))
        this.start(pad, running)
        break
      case 'stopMultiPad':
        if (this.has(pad)) this.setLamp(pad, 'ready')
        break
      case 'stopAllMultiPads':
        st.pads.forEach((p, i) => this.has(i) && this.setLamp(i, 'ready'))
        break
      case 'armMultiPad':
        if (this.has(pad)) {
          const l = st.pads[pad].lamp
          if (l === 'armed') this.setLamp(pad, 'ready')
          else if (l === 'ready') this.setLamp(pad, 'armed')
        }
        break
      case 'setMultiPadRepeat':
        if (this.has(pad)) st.pads[pad].repeat = c.on
        break
      case 'setMultiPadChordMatch':
        if (this.has(pad)) st.pads[pad].chordMatch = c.on
        break
      case 'setMultiPadSynchroStop':
        st.synchroStop = { styleStop: c.styleStop, ending: c.ending }
        break
    }
    return null
  }

  /** The band moved on `beats` beats (or the clock did while stopped). */
  beats(beats: number) {
    this.st.pads.forEach((p, i) => {
      if (p.lamp !== 'playing') return
      this.left[i] -= beats
      if (this.left[i] <= 0) {
        if (p.repeat) this.left[i] += 4
        else p.lamp = 'ready'
      }
    })
  }

  /** A bar line while the band plays: queued pads start. */
  bar() {
    this.st.pads.forEach((p, i) => p.lamp === 'queued' && this.setLamp(i, 'playing'))
  }

  /** The band started: armed pads start with it. */
  bandStarted() {
    this.st.pads.forEach((p, i) => p.lamp === 'armed' && this.setLamp(i, 'playing'))
  }

  /** A chord in the chord section: armed pads start. */
  chord(running: boolean) {
    this.st.pads.forEach((p, i) => p.lamp === 'armed' && this.start(i, running))
  }

  private stopRepeating() {
    this.st.pads.forEach((p, i) => p.repeat && (p.lamp === 'playing' || p.lamp === 'queued') && this.setLamp(i, 'ready'))
  }

  bandStopped() {
    if (this.st.synchroStop.styleStop) this.stopRepeating()
  }

  endingStarted() {
    if (this.st.synchroStop.ending) this.stopRepeating()
  }

  panic() {
    this.st.pads.forEach((p, i) => this.has(i) && this.setLamp(i, 'ready'))
  }
}

File: app/src-tauri/src/mock_multipad.rs
//! The mock's Multi Pads (the twin of `app/src/lib/api/mock-multipad.ts`): the engine's
//! behaviour (docs/multipad.md), closely enough for the Multi Pad panel. No audio.

use yahaha::api::{MultiPadBank, MultiPadBankEntry, MultiPadCmd, MultiPadPad, MultiPadState, MultiPadSynchroStop, PadLamp};

const ROOT: &str = "/Users/me/Styles";

/// (name, folder, pad names, repeat, chord match): the synthetic demo bank and two more.
type Bank = (&'static str, &'static str, [&'static str; 4], [bool; 4], [bool; 4]);
const BANKS: [Bank; 3] = [
    ("Demo", "Pads", ["Shaker Loop", "Rise Arp", "Bass Riff", "Brass Hit"], [true, false, true, false], [false, true, true, true]),
    ("Latin Perc", "Pads/Latin", ["Conga Loop", "Timbale Fill", "Cowbell", ""], [true, false, true, false], [false; 4]),
    ("Strings FX", "Pads", ["Swell", "Pizz Run", "Stab", "Tremolo"], [false, false, false, true], [true; 4]),
];

fn empty_pads() -> Vec<MultiPadPad> {
    (0..4u8).map(|i| MultiPadPad { index: i, channel: 5 + i, ..MultiPadPad::default() }).collect()
}

pub fn initial() -> MultiPadState {
    MultiPadState {
        bank: None,
        loading: false,
        pads: empty_pads(),
        synchro_stop: MultiPadSynchroStop::default(),
        banks: BANKS
            .iter()
            .enumerate()
            .map(|(id, b)| MultiPadBankEntry { id, name: b.0.into(), folder: b.1.into(), path: format!("{ROOT}/{}/{}.pad", b.1, b.0) })
            .collect(),
    }
}

/// Beats left of each playing pad's pass (every mock pad is one 4-beat bar).
#[derive(Default)]
pub struct MockPads {
    left: [f64; 4],
}

impl MockPads {
    fn set(&mut self, st: &mut MultiPadState, i: usize, lamp: PadLamp) {
        st.pads[i].lamp = lamp;
        if lamp == PadLamp::Playing {
            self.left[i] = 4.0;
        }
    }

    fn has(st: &MultiPadState, i: usize) -> bool {
        st.pads.get(i).is_some_and(|p| p.lamp != PadLamp::Empty)
    }

    fn start(&mut self, st: &mut MultiPadState, i: usize, running: bool) {
        if Self::has(st, i) {
            self.set(st, i, if running { PadLamp::Queued } else { PadLamp::Playing });
        }
    }

    fn each(&mut self, st: &mut MultiPadState, from: PadLamp, to: impl Fn(&MultiPadPad) -> Option<PadLamp>) {
        for i in 0..4 {
            if st.pads[i].lamp == from
                && let Some(l) = to(&st.pads[i])
            {
                self.set(st, i, l);
            }
        }
    }

    /// A pad command; an error message for a refused one.
    pub fn cmd(&mut self, st: &mut MultiPadState, c: MultiPadCmd, running: bool) -> Option<String> {
        let pad = match &c {
            MultiPadCmd::TriggerMultiPad { pad }
            | MultiPadCmd::StopMultiPad { pad }
            | MultiPadCmd::ArmMultiPad { pad }
            | MultiPadCmd::SetMultiPadRepeat { pad, .. }
            | MultiPadCmd::SetMultiPadChordMatch { pad, .. } => Some(*pad as usize),
            _ => None,
        };
        if let Some(p) = pad.filter(|&p| p > 3) {
            return Some(format!("no Multi Pad {p} (pads are 0-3)"));
        }
        let i = pad.unwrap_or(0);
        match c {
            MultiPadCmd::LoadMultiPad { id } => return self.load(st, id).err(),
            MultiPadCmd::LoadMultiPadPath { path } => match st.banks.iter().position(|b| b.path == path) {
                Some(id) => return self.load(st, id).err(),
                None => return Some(format!("{path}: not a MIDI/Multi Pad file")),
            },
            MultiPadCmd::ClearMultiPad => {
                st.bank = None;
                st.pads = empty_pads();
            }
            MultiPadCmd::TriggerMultiPad { .. } => {
                for k in 0..4 {
                    if st.pads[k].lamp == PadLamp::Armed {
                        self.start(st, k, running);
                    }
                }
                self.start(st, i, running);
            }
            MultiPadCmd::StopMultiPad { .. } => {
                if Self::has(st, i) {
                    self.set(st, i, PadLamp::Ready);
                }
            }
            MultiPadCmd::StopAllMultiPads => self.panic(st),
            MultiPadCmd::ArmMultiPad { .. } => match st.pads[i].lamp {
                PadLamp::Armed => self.set(st, i, PadLamp::Ready),
                PadLamp::Ready => self.set(st, i, PadLamp::Armed),
                _ => {}
            },
            MultiPadCmd::SetMultiPadRepeat { on, .. } => {
                if Self::has(st, i) {
                    st.pads[i].repeat = on;
                }
            }
            MultiPadCmd::SetMultiPadChordMatch { on, .. } => {
                if Self::has(st, i) {
                    st.pads[i].chord_match = on;
                }
            }
            MultiPadCmd::SetMultiPadSynchroStop { style_stop, ending } => st.synchro_stop = MultiPadSynchroStop { style_stop, ending },
        }
        None
    }

    fn load(&mut self, st: &mut MultiPadState, id: usize) -> Result<(), String> {
        let Some(b) = BANKS.get(id) else { return Err(format!("no Multi Pad bank {id}")) };
        st.bank = Some(MultiPadBank { id, name: b.0.into(), path: st.banks[id].path.clone() });
        st.pads = (0..4)
            .map(|i| {
                let has = !b.2[i].is_empty();
                MultiPadPad {
                    index: i as u8,
                    name: b.2[i].into(),
                    lamp: if has { PadLamp::Ready } else { PadLamp::Empty },
                    repeat: has && b.3[i],
                    chord_match: has && b.4[i],
                    channel: 5 + i as u8,
                }
            })
            .collect();
        Ok(())
    }

    /// The clock moved on `beats` beats.
    pub fn beats(&mut self, st: &mut MultiPadState, beats: f64) {
        for i in 0..4 {
            if st.pads[i].lamp != PadLamp::Playing {
                continue;
            }
            self.left[i] -= beats;
            if self.left[i] <= 0.0 {
                if st.pads[i].repeat {
                    self.left[i] += 4.0;
                } else {
                    st.pads[i].lamp = PadLamp::Ready;
                }
            }
        }
    }

    /// A bar line while the band plays: queued pads start.
    pub fn bar(&mut self, st: &mut MultiPadState) {
        self.each(st, PadLamp::Queued, |_| Some(PadLamp::Playing));
    }

    pub fn band_started(&mut self, st: &mut MultiPadState) {
        self.each(st, PadLamp::Armed, |_| Some(PadLamp::Playing));
    }

    pub fn chord(&mut self, st: &mut MultiPadState, running: bool) {
        let to = if running { PadLamp::Queued } else { PadLamp::Playing };
        self.each(st, PadLamp::Armed, |_| Some(to));
    }

    fn stop_repeating(&mut self, st: &mut MultiPadState) {
        for from in [PadLamp::Playing, PadLamp::Queued] {
            self.each(st, from, |p| p.repeat.then_some(PadLamp::Ready));
        }
    }

    pub fn band_stopped(&mut self, st: &mut MultiPadState) {
        if st.synchro_stop.style_stop {
            self.stop_repeating(st);
        }
    }

    pub fn ending_started(&mut self, st: &mut MultiPadState) {
        if st.synchro_stop.ending {
            self.stop_repeating(st);
        }
    }

    pub fn panic(&mut self, st: &mut MultiPadState) {
        for i in 0..4 {
            if Self::has(st, i) {
                self.set(st, i, PadLamp::Ready);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn load_press_queue_and_synchro_stop() {
        let (mut st, mut p) = (initial(), MockPads::default());
        assert!(p.cmd(&mut st, MultiPadCmd::LoadMultiPad { id: 0 }, false).is_none());
        assert_eq!(st.bank.as_ref().map(|b| b.name.as_str()), Some("Demo"));
        p.cmd(&mut st, MultiPadCmd::TriggerMultiPad { pad: 2 }, true);
        assert_eq!(st.pads[2].lamp, PadLamp::Queued);
        p.bar(&mut st);
        assert_eq!(st.pads[2].lamp, PadLamp::Playing);
        p.band_stopped(&mut st);
        assert_eq!(st.pads[2].lamp, PadLamp::Ready);
        assert!(p.cmd(&mut st, MultiPadCmd::ArmMultiPad { pad: 9 }, false).is_some());
    }
}
